### PR TITLE
Test environment

### PR DESCRIPTION
**Pull request by alexsavio**
Removed the 'manager' VM, useless now.
Now using the Vagrant `ansible` provisioner.

### CODE REVIEW

**Review by hardion — COMMENTED** [review]
You can remove the section vagrant manager from the README

File: ansible.cfg [hunk 1]
```diff
@@ -1,450 +1,9 @@
...
-# running in background processes to fail.  Uncomment this line to have
-# Paramiko automatically add host keys.
-#host_key_auto_add = True
-
-[ssh_connection]
-
-# ssh arguments to use
-# Leaving off ControlPersist will result in poor performance, so use
-# paramiko on older platforms rather than removing it, -C controls compression use
-ssh_args = -o ControlMaster=auto -o ControlPersist=60s -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes
```
[the remark this answers is not in the document]
Comment (alexsavio) [reply ansible.cfg]: I added another setting recommended by @artcz: `ssh_args = -o ControlMaster=auto -o ControlPersist=60s -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes`

[the remark this answers is not in the document]
Comment (hardion) [reply ansible.cfg]: OK the ssh_args line is at the bottom of the diff.

[the remark this answers is not in the document]
Comment (alexsavio) [reply ansible.cfg]: oh, it's the same?

[the remark this answers is not in the document]
Comment (hardion) [reply ansible.cfg]: A bit different the cache has been set to 30 min instead of 60s.

File: roles/common/tasks/main.yml [hunk 2]
```diff
@@ -17,6 +17,17 @@
...
-  become: yes
-  become_method: sudo
   tags: always
+
+- name: Install dependencies
+  apt:
+    name: '{{ item }}'
+    update_cache: yes
+    cache_valid_time: 3600
+  with_items:
```
Comment (webknjaz) [inline roles/common/tasks/main.yml]: Move this to name. There's no need to create 6 subtasks, apt module knows how to deal with a list.

File: ansible.cfg (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexsavio) [reply ansible.cfg]: Not sure what value to put there.

[the remark this answers is not in the document]
Comment (artcz) [reply ansible.cfg]: I guess either one is fine for local. For remote I think the higher the value the less ssh connections it has to establish, but that might not be true.

[the remark this answers is not in the document]
Comment (alexsavio) [reply ansible.cfg]: thanks

**Comment by alexsavio** [comment]
Thanks for the reviews @webknjaz, @hardion, and @artcz

File: roles/docker-compose/tasks/install_docker.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    repo: deb [arch=amd64] https://download.docker.com/linux/ubuntu {{ansible_distribution_release}} stable
+    state: present
+    update_cache: yes
+
+- name:
+  apt:
+    name: '{{ item }}'
+    update_cache: yes
+    cache_valid_time: 3600
+  with_items:
```
Comment (webknjaz) [inline roles/docker-compose/tasks/install_docker.yml]: This is probably unnecessary for one-item lists + should be moved to name anyway

File: roles/docker-compose/tasks/install_docker_compose.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    state: forcereinstall
+  with_items:
+    - pipsi
+  executable: pip3
+  become: yes
+  become_method: sudo
+
+- name: Install docker-compose
+  command:
+    pipsi install docker-compose==1.22.0
```
Comment (webknjaz) [inline roles/docker-compose/tasks/install_docker_compose.yml]: While this works, the recommend approach is to write a module wrapping this.

File: roles/python/tasks/main.yml [hunk 5]
```diff
@@ -2,24 +2,11 @@
 
 - name: Install some apt packages
   apt:
-    name: '{{ item }}'
-  with_items:
-    - libyaml-dev
+    name: '{{ packages }}'
```
Comment (webknjaz) [inline roles/python/tasks/main.yml]: Too much complexity here. Just list packages under name directly.

File: vagrant/test.sh [hunk 6]
```diff
@@ -1,11 +1,13 @@
 #!/bin/bash
 
+set -eo pipefail
```
Comment (webknjaz) [inline vagrant/test.sh]: Also: `-u` - prevents from using undeclared vars

File: vagrant/test.sh [hunk 7]
```diff
@@ -1,11 +1,13 @@
 #!/bin/bash
```
Comment (webknjaz) [inline vagrant/test.sh]: Why didn't you ansiblize this script?

File: roles/python/tasks/main.yml (hunk 5 above)
Comment (alexsavio) [reply roles/python/tasks/main.yml]: Do you mean:
```
    name:
        - package1
        - package2
```

?

File: roles/docker-compose/tasks/install_docker_compose.yml (hunk 4 above)
Comment (alexsavio) [reply roles/docker-compose/tasks/install_docker_compose.yml]: How do I write a module to wrap this?

File: roles/python/tasks/main.yml (hunk 5 above)
Comment (webknjaz) [reply roles/python/tasks/main.yml]: Yep. I would understand if you wanted to put them into a dedicated place somewhere else. But declaring vars in a task to reuse them one line above is weird.

File: vagrant/test.sh (hunk 7 above)
Comment (webknjaz) [reply vagrant/test.sh]: :+1:

File: roles/python/tasks/main.yml (hunk 5 above)
Comment (alexsavio) [reply roles/python/tasks/main.yml]: I took it from the example in the Ansible documentation: https://docs.ansible.com/ansible/latest/modules/apt_module.html#examples

Comment (alexsavio) [reply roles/python/tasks/main.yml]: But it looks better your way, I just wasn't sure what was the YAML syntax for that

File: roles/docker-compose/tasks/install_docker_compose.yml (hunk 4 above)
Comment (webknjaz) [reply roles/docker-compose/tasks/install_docker_compose.yml]: It's relatively simple. You may look up some of the existing modules:
* https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/os/portage.py
* https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/language/pip.py
* https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/os/apt.py

Personally, I prefer OOP approach (wrapping that into a class with entry point and custom exception machinery), like in:
https://github.com/webknjaz/ansible-gentoo-laptop/blob/master/library/eix.py
(yes, if you put it into `library/` folder in this repo, you'll be able to use it by its name)

There's `ansible.module_utils.basic.AnsibleModule` helper you can import if you're writing it in Python.

Some docs on this are here: https://docs.ansible.com/ansible/latest/dev_guide/developing_modules.html

*TL;DR* your module accepts some JSON via stdin and responds with JSON to stdout. That's how ansible core is talking to modules.
Initialize `AnsibleModule` helper containing schema for accepted args. Then this instance has `.exit_json()` and `.fail_json()` to return some result. Other parts of flow are up to you, in general you'd run some tool you're wrapping via subprocess and then check the result.
Keep in mind to use Python syntax compatible with the interpreter on target hosts, best thing to do is useing subset which is common across Python 2 and Python 3 + also have `__future__` imports + `__metaclass__ = type` hack (makes classes in Python 2 new-style by default, no need to inherit from object explicitly).

**Comment by webknjaz** [comment]
> Thanks for the reviews @webknjaz, @hardion, and @artcz

FYI if you want to make GitHub aware of someone helping you with code, you can add trailers to the commit description, like:

`Co-authored-by: Name Surname <email.bound@to.their.github.com>`

File: roles/python/tasks/main.yml (hunk 5 above)
Comment (webknjaz) [reply roles/python/tasks/main.yml]: Oh, our docs probably need updating :)

In general, YAML knows JSON-like syntax for sequences `name: [package1, package2]` and human-readable:
```yaml
name:
- package1
- package2
```
I prefer to not do indentation because it already knows that it's a sequence from the `-` sign.

When you use `name: "{{ some_var }}"` Ansible renders it as a Jinja2 template and preserves type, so that it's actually a list, not a string :)

File: roles/docker-compose/tasks/install_docker_compose.yml (hunk 4 above)
Comment (alexsavio) [reply roles/docker-compose/tasks/install_docker_compose.yml]: nice! thanks. you should make a blog post with just this :smile:

Comment (webknjaz) [reply roles/docker-compose/tasks/install_docker_compose.yml]: I think, you can find some existing blog posts in the wild :)
